[NVPTX] Add Volta Atomic SequentiallyConsistent Load and Store Operations

gonzalobg opened the pull request:
This PR Builds on #98022 .
It adds support for Volta's SequentiallyConsistent Load and Store operations at system scope.

gonzalobg committed:
[NVPTX] Volta SequentiallyConsistent Load/Store Ops

gonzalobg committed:
[NVPTX] Align Memory Ordering enum with LLVM

gonzalobg committed:
[NVPTX]: Fix typos

gonzalobg committed:
[NVPTX] Refactor and clean up load,tryLoad,tryStore a bit

gonzalobg commented:
@AlexMaclean @Artem-B @jlebar hello all three, I'm looking for some reviewers for this series of PRs to improve the support of atomics in LLVM NVPTX backend. It's my first time touching these parts of the LLVM project, but everyone in the LLVM Discord has been super nice helping me get onboarded. 

This PR includes all the changes of #96436  and #98022, which haven't really gotten any replies yet.. I just kept building on them. Not sure whether I should close them and continue building on them, or whether it is easier to review them piece meal.

I'm happy to help review other people's PR and help with the NVPTX backend in general. I'm "gonzalob" in the LLVM Discord in case anyone wants to chat :)

Artem-B commented:
Sorry, I was out of office for a while. I'm catching up on the pending reviews now. Stay tuned.

gonzalobg committed:
[NVPTX] Support fence instruction

gonzalobg committed:
[NVPTX] Cleanup SeqCst Load/Store

gonzalobg committed:
[NVPTX] Add vector load/store tests and refactor load/store tests

gonzalobg committed:
[NVPTX] Cleanups
- Remove dead code in InstPrinter
- Capitalization and improve Table comments
- Update fence->membar lowering

gonzalobg commented:
Have rebased this on top of https://github.com/llvm/llvm-project/pull/99649 . This is now a simpler change.

gonzalobg committed:
[NVPTX] Update atomic volatile unordered test

gonzalobg committed:
[NVPTX] Update comment: now handling Unordered

gonzalobg commented:
Have rebased on top of the merged PRs, and extended the `unordered` tests in the same way this PR was extending the other tests.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -860,26 +887,51 @@ static unsigned int getCodeMemorySemantic(MemSDNode *N,
+      OS << "NVPTX does not support SequentiallyConsistent Ordering on "
+            "read-modify-writes yet: "
+         << N->getOperationName();
+      N->print(OS);
```
We could use a utility function returning `std::string` representation of the node, and then just use the plain strings to construct the error message. This may be done in a separate patch.

```
std::string repr(const SDNode &N) {
      SmallString<256> Msg;
      raw_svector_ostream OS(Msg);
      N->print(OS);
      return OS.str();
}
...
report_fatal_error("some error message about: " + repr(N));
```

gonzalobg replied:
I ended up using `llvm::formatv`, see the "Cleanup error reporting commit".
Its not precisely what you had in mind here, let me know if its better (there may be something worth refactoring there as well).

Artem-B replied:
formatv works. Definitely cleaner than a piecemeal message construction via a string stream.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store-sm-70.ll:
```diff
@@ -1,169 +1,7 @@
 ; RUN: llc < %s -march=nvptx64 -mcpu=sm_70 -mattr=+ptx82 | FileCheck %s
 ; RUN: %if ptxas-12.2 %{ llc < %s -march=nvptx64 -mcpu=sm_70 -mattr=+ptx82 | %ptxas-verify -arch=sm_70 %}
 
-; CHECK-LABEL: generic_plain
```
Why were these removed? I think it still makes sense to make sure that plain non-atomic loads/stores still work as expected. It may be redundant with other load/store tests. In that case please add a comment 'plain loads/stores without atomic modifiers are tested in ....'

Also, if the file is now testing only atomic loads/stores, it would make sense to rename it accordingly. Then the comment about non-atomic variants will not be necessary.

gonzalobg replied:
> Why were these removed? I think it still makes sense to make sure that plain non-atomic loads/stores still work as expected. It may be redundant with other load/store tests. In that case please add a comment 'plain loads/stores without atomic modifiers are tested in ....'

- `load-store.ll` contains all tests that are supported by all sms, covering sm20 and sm70 using different checks. 
- `load-store-sm70.ll` only contains the ones that require sm70+.

The generic plain tests in load-store-sm70.ll were duplicates from load-store.ll. They were testing different codegen, but the test was the same, so now they are covered there via different checks. 

Does that make sense? 

I'll add comments.

> Also, if the file is now testing only atomic loads/stores, it would make sense to rename it accordingly. 

Right now they only cover atomic loads/stores, but they will cover other new ops eventually, like for example `.b128` weak (non-atomic) loads/stores, and newer ops that were made available later. 

Once we add `syncscope` support, there will also be a `-sm-90` version of these tests to cover things like `cluster` scope.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -1,5 +1,7 @@
-; RUN: llc < %s -march=nvptx64 -mcpu=sm_20 | FileCheck %s
+; RUN: llc < %s -march=nvptx64 -mcpu=sm_20 | FileCheck -check-prefixes=CHECK,SM60 %s
```
There's a bit of a contradiction between `-mcpu=sm_20` and `-check-prefixes=SM60`.
Perhaps rename the check `NOATOMIC` ?

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -578,47 +1011,134 @@ define void @shared_volatile(ptr addrspace(3) %a, ptr addrspace(3) %b, ptr addrs
+  ; Note: per https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#vectors
+  ; vectors cannot exceed 128-bit in length, i.e., .v4.u64 is not allowed.
+
+  ; TODO: should be combined into single .u64 op
```
Why? If we're talking specifically about v2.f32, loading them ad u64 would not give us any benefits, would it? If anything, we'll need to  generate a more verbose PTX to move from 64-bit integer register into a pair of f32, so we could add another FP value, and then glue and convert them back to u64.

AFAICT the comments here and above would apply only to the vectors exceeding 4 elements, but still under 128bits in total size, but none of the tests fall into that category. 

I would suggest removing the comments for now, or add some extra test cases where they would indicate a furute optimization opportunity. As things stand, I do not see any practical benefits for switching to sngle-element loads/stores for vectors that are 4 elements or less in size. Perhaps I'm missing something.

gonzalobg replied:
> Why?

In a nutshell, it mostly only matters for testing. First, vectors only matter for atomics: they enable PTX to use smaller atomic ops if profitable (a larger scalar atomic op forbids that optimization). For scalars, vector vs scalar does not matter. The hardware only supports scalars, so if we provide a scalar, PTX lowers it in the straight forward way to hardware, and if we provide a vector, PTX will combine the registers into a scalar, and then issue a scalar op. 

For testing, generating scalar vs vector ops depending on whether mid-level LLVM optimizations generated one or the other is not great (if we want to generate these tests with table gen, which will continue to explode in size as we add scopes and other features, that bump the cartesian product of options). 

We can canonicalize all of these in the NVPTX backend to generate some canonical PTX. If we pick scalars, we help PTX a little bit, and the canonicalization is unique (from scalar and different vector types to the same scalar). We could pick other canonicalizations (e.g. to some vector type, but then u32 -> v4u8 is weird, and whether we should pick v4u8 or v2u16 is also weird).

So i think that for weak ops, we should just canonicalize to an unsigned integer (in a similar way that we currently do not preserve "signedness" from the last op of a register into whether we emit a signed s32 vs unsigned u32 load/store op; something that also doesn't matter).

So this was more or less my thinking. What do you think?

Artem-B replied:
> Right now, it seems that we canonicalize a vector to scalars in some cases (e.g. <i8 x 4> to .u32 ops), but not others (e.g. <i16 x 2> lowers to v2u16 ops).

> Do you know why?

Probably due to https://github.com/llvm/llvm-project/pull/67866 handling v4i8 as a special case.

gonzalobg replied:
I see, makes sense. I've replaces the TODOs with a top level one that just says that we are treating these inconsistently and that's it. What's the best way to treat these here is not clear to me.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -578,47 +1011,134 @@ define void @shared_volatile(ptr addrspace(3) %a, ptr addrspace(3) %b, ptr addrs
+  ; CHECK: st.volatile.shared.v2.u32 [%rd{{[0-9]+}}], {%r{{[0-9]+}}, %r{{[0-9]+}}}
+  store volatile <2 x i32> %l.add, ptr addrspace(3) %d
+
+  ; TODO: should be combined into single .b128 op in sm_70+
```
It's a bit of an off-topic for the patch, but I'd be curious to know what it would buy us.
AFAICT, on sm_70, the hardware still uses 32-bit registers, so a .b128 load will work exactly the same as `v4.b32` load, only the latter will not need an additional move splitting the 128-bit chunk into 32-bit bits.

gonzalobg replied:
Have redone the comments throughout to be clearer.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -826,11 +1439,84 @@ define void @local_volatile(ptr addrspace(5) %a, ptr addrspace(5) %b, ptr addrsp
 
 ; CHECK-LABEL: local_monotonic
 define void @local_monotonic(ptr addrspace(5) %a, ptr addrspace(5) %b, ptr addrspace(5) %c, ptr addrspace(5) %d, ptr addrspace(5) %e) local_unnamed_addr {
+  ; TODO: this codegen looses Concurrent Forward Progress
```
This comment needs more context. Can you elaborate on what you mean?

gonzalobg replied:
Have redone the comments throughout to be clearer.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -867,11 +1553,15 @@ define void @local_monotonic(ptr addrspace(5) %a, ptr addrspace(5) %b, ptr addrs
   ; CHECK: st.local.f64 [%rd{{[0-9]+}}], %fd{{[0-9]+}}
   store atomic double %f.add, ptr addrspace(5) %e monotonic, align 8
 
+  ; TODO: LLVM IR Verifier does not support atomics on vector types.
```
Rephrase in terms what we need to fix or do. E.g. `add atomic vector loads/stores when LLVM verifier can deal with them`. Bonus point for including commented out sample of IR which triggers that verifier failure, so whoever has questions can easily reproduce it.

gonzalobg replied:
Have redone the comments throughout to be clearer.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -990,5 +1684,10 @@ define void @local_unordered_volatile(ptr addrspace(5) %a, ptr addrspace(5) %b,
   ret void
 }
+
+; TODO: missing .const statespace tests
```
Nit: `add missing <X/Y/Z> tests...`

gonzalobg replied:
Have redone the comments throughout to be clearer.

Artem-B reviewed (commented):
Looks OK overall, with some minor comments.

I'll be out till Tuesday, and will not be able to do the next review round until later next week.

gonzalobg committed:
[NVPTX] refactor NVPTX::Ordering to string/stream

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTX.h:
```diff
@@ -106,15 +106,53 @@ enum LoadStore {
+  LAST = RelaxedMMIO
 };
+
+template <typename OStream> OStream &operator<<(OStream &O, Ordering Order) {
```
Good idea. I've done that split into a "to c string" (`char const*`) conversion plus using `raw_ostream`.
These are inline functions here for now, but we could move their implementation to a .cpp file. Itried a few of them, and couldn't find one that still allows everything to link. Do you have a suggestion for which .cpp file to put these in? Or should I just leave them as inline functions here for now?

Artem-B replied:
Moving them to NVPTXUtility.* would probably make sense.

gonzalobg replied:
I tried that but it failed to link throughout. I think I'd need to increase the set of TUs that link against it. Would it be possible for me to do it in a separate PR? 

The next PR in the queue is a NFC refactor, so it would fit in there.

gonzalobg replied:
Fixed that here, resolving.

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -714,21 +714,28 @@ static unsigned int getCodeAddrSpace(MemSDNode *N) {
 
-static unsigned int getCodeMemorySemantic(MemSDNode *N,
-                                          const NVPTXSubtarget *Subtarget) {
+struct OperationOrderings {
```
Good point. Have moved these two into an anonymous namespace to avoid collisions.

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -860,26 +887,51 @@ static unsigned int getCodeMemorySemantic(MemSDNode *N,
+    // This sets the ordering of the fence to SequentiallyConsistent, and
+    // sets the corresponding ordering for the instruction.
+    NVPTX::Ordering InstrOrder;
+    if (N->readMem()) {
```
Thanks, weird that clang-format did not catch this one :(

gonzalobg committed:
[NVPTX] Refactor OperationOrderings into anonymous namespace to avoid collisions

gonzalobg committed:
[NVPTX] Cleanup error reporting

gonzalobg committed:
[NVPTX] Cleanup comments in tests

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -82,45 +153,139 @@ define void @generic_volatile(ptr %a, ptr %b, ptr %c, ptr %d) local_unnamed_addr
+  ; - provide support for vector atomics, e.g., by extending LLVM IR or via intrinsics
+  ; - update tests in load-store-sm70.ll as well.
+
+  ; TODO: make this operation consistent with the one for <4 x i8>
```
This may be due to the fact that a while back I've plumbed through <4 x i8> lowering as a special case, but left other combinations untouched and they still fall back to the default lowering.

gonzalobg replied:
Thanks for the hint and for the link to that PR. 
I don't know what's the best way to address this, but will explore and will let you know.

Artem-B replied:
It's not something that needs fixing right away, I think. In any case, it's unrelated to atomics and should be done in a separate patch.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -82,45 +153,139 @@ define void @generic_volatile(ptr %a, ptr %b, ptr %c, ptr %d) local_unnamed_addr
+  ; TODO: make this operation consistent with the one for <2 x i8>
+  ; This operation lowers to a "full vector volatile PTX operation".
+  ; CHECK: ld.volatile.u32 %r{{[0-9]+}}, [%rd{{[0-9]+}}]
+  %i.load = load volatile <4 x i8>, ptr %c
```
We probably want `<8 x i8>`, and, maybe, `<6 x i8>` too.  i8 in PTX is a really odd type.

Speaking of odd types, we should also test `i1`.

gonzalobg replied:
Agreed. Would it be ok for me to leave a TODO here for those, and try to do that as part of a NFC ? (if these types break, chances are a bit more work will be required to fix that, which may be independent to this).

Artem-B replied:
Separate patch is fine.

gonzalobg replied:
Moved to a comment.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -289,6 +466,69 @@ define void @global_plain(ptr addrspace(1) %a, ptr addrspace(1) %b, ptr addrspac
+  store <2 x i64> %n.add, ptr addrspace(1) %d
+
+  ; Note: per https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#vectors
+  ; vectors cannot exceed 128-bit in length, i.e., .v4.u64 is not allowed.
```
That sounds like a good reason to verify that we do something sensible when someone does try to do an atomic op on a `<4 x i64>`.

gonzalobg replied:
Yeah, I've learned that we can add tests that fail. 
I think this would be a good use for that.
I'm going to turn this into a TODO for now, instead of a note.

Artem-B commented on llvm/test/CodeGen/NVPTX/load-store.ll:
```diff
@@ -785,11 +1290,77 @@ define void @local_plain(ptr addrspace(5) %a, ptr addrspace(5) %b, ptr addrspace
 
 ; CHECK-LABEL: local_volatile
 define void @local_volatile(ptr addrspace(5) %a, ptr addrspace(5) %b, ptr addrspace(5) %c, ptr addrspace(5) %d) local_unnamed_addr {
+  ; TODO: generate PTX that preserves Concurrent Forward Progress
+  ;       by using volatile operations.
```
I'm still not sure what the `PTX that preserves Concurrent Forward Progress` would look like.

Is the comment implying that LLVM produces invalid PTX right now (i.e. not preserving concurrent forward progress)?
Or does it mean that we're just missing a test case which would demonstrate that we're generating valid PTX for that use case?

gonzalobg replied:
It means that we are producing invalid PTX right now in some corner cases that very few people run into.

There are a few ways to generate correct PTX here, but the gist of it is that we need to preserve volatile or atomic in the PTX, we can't just drop it and issue weak ops.

One option is, e.g., generating precisely the code that we are generating right now, but with an additional volatile/atomic load whose result is unused (its dead) from some dummy memory location.

I'll work on this eventually but we'll need to add a few more tests (to .const, and .param) first, because those suffer of a similar issue, and when we try to fix this we should make sure we fix this throughout.

Artem-B replied:
It would be useful to open an issue for that, so it does not slip through the cracks.

gonzalobg replied:
Have filled: https://github.com/llvm/llvm-project/issues/101429

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTX.h:
```diff
@@ -106,15 +106,49 @@ enum LoadStore {
-  Release = 4,
-  RelaxedMMIO = 5
+  // Consume = 3,   // Unimplemented in LLVM; NVPTX would map to "Acquire"
+  Acquire = 4,
```
I don't think we really really need them, but in the DAG they are just `(i32 x)` nodes where `x` is one of these values, and the values picked here match the ones of LLVM `AtomicOrdering` for the cases in which both expose the same ordering, which makes it easy to interpret those DAG nodes.

gonzalobg replied:
Good idea with the `static_assert`s, that would prevent this from breaking.
Since we have some orderings that extend AtomicOrdering and we have to set `LAST`, etc., I've left the integer values though. If a `static_assert` fail, we'll need to update this properly.

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -1889,7 +1962,25 @@ bool NVPTXDAGToDAGISel::tryStore(SDNode *N) {
 
   // Memory Semantic Setting
-  unsigned int CodeMemorySem = getCodeMemorySemantic(ST, Subtarget);
+  auto [InstructionOrdering, FenceOrdering] =
```
Not fully sure what you mean. The code is in a function that returns a struct. This `auto [x, y] = value;` is just extracting the fields of the struct into separate variables.

I can also just do:

```cpp=
auto Orderings = getOperationOrdering(...);
auto InstructionOrdering = Orderings.InstructionOrdering;
auto FenceOrdering = Orderings.FenceOrdering;
```

?

Artem-B replied:
I mean the whole block of code from here to the end of the switch below.

gonzalobg replied:
Thanks, have moved this into a helper function :)

gonzalobg committed:
[NVPTX] Simplify NVPTX::Ordering by not making it an enum class

gonzalobg committed:
[NVPTX] Move Ordering to string functions to utilities

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -714,21 +715,30 @@ static unsigned int getCodeAddrSpace(MemSDNode *N) {
+namespace {
+
+struct OperationOrderings {
+  NVPTX::OrderingUnderlyingType InstrOrdering;
+  NVPTX::OrderingUnderlyingType FenceOrdering;
```
Fixed

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTX.h:
```diff
@@ -106,15 +106,49 @@ enum LoadStore {
+  LAST = RelaxedMMIO
 };
+
+inline char const *toCString(Ordering Order) {
```
Ended up moving them to .h. Found the issue with the .cpp: the AsmPrinter in MCTargetDesc needs these, but NVPTXUtilities depends on it (because NVPTX depends on it), so we end up with a cyclic dependency. Punting moving these to the .cpp file for now..

gonzalobg committed:
[NVPTX] Improve comments in load-store tests

gonzalobg committed:
[NVPTX] Refactor fence insertion for loads/stores

gonzalobg committed:
[NVPTX] Refactor memory ordering

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -826,62 +853,74 @@ static unsigned int getCodeMemorySemantic(MemSDNode *N,
+          formatv("NVPTX does not support SequentiallyConsistent Ordering on "
+                  "read-modify-writes yet: {}",
+                  N->getOperationName()));
+    return AddrGenericOrGlobalOrShared
```
Indeed, this was historical from doing this piece meal. Have moved this to the top.

Artem-B reviewed (commented):
Few more minor nits, and I think we should be good to go.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -820,68 +845,76 @@ static unsigned int getCodeMemorySemantic(MemSDNode *N,
        CodeAddrSpace == NVPTX::PTXLdStInstCode::GLOBAL ||
        CodeAddrSpace == NVPTX::PTXLdStInstCode::SHARED);
+  if (!AddrGenericOrGlobalOrShared)
+    return NVPTX::Ordering::NotAtomic;
```
Are there any cases we may want to use volatile here? 

We probably do not need it for 'local'. But I'm not quite sure about `.const`. Technically it could be updated by the host and we may need to have a way to force a read from it.

gonzalobg replied:
Right now, according to the PTX spec, doing a `volatile` of a `.const` is undefined behavior. 
We should probably fix that in the spec, the PTX implementation, and once we add the `.const` and `.param` tests, update all of this as needed.

gonzalobg replied:
> Technically it could be updated by the host and we may need to have a way to force a read from it.

If the host updates a `const` while a kernel that has accessed it once has not finished, the behavior is undefined.

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXUtilities.h:
```diff
@@ -82,6 +83,36 @@ inline unsigned promoteScalarArgumentSize(unsigned size) {
+
+namespace NVPTX {
+
+inline char const *toCString(Ordering Order) {
```
Have changed the name, and the return type changed to `string_view`.

gonzalobg replied:
The lifetime of `char const*` for these string is "static lifetime" / "whole program" right?

gonzalobg commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -1186,36 +1183,29 @@ bool NVPTXDAGToDAGISel::tryLoad(SDNode *N) {
-                     Offset,
-                     Chain};
-    NVPTXLD = CurDAG->getMachineNode(*Opcode, dl, TargetVT, MVT::Other, Ops);
+    Ops.push_back(Base);
```
Thanks, didn't knew `SmallVector` had this!

gonzalobg committed:
[NVPTX] Refactor and cleanups

gonzalobg committed:
[NVPTX] Switch to std::string

gonzalobg commented:
@Artem-B This is green!

Artem-B commented on llvm/lib/Target/NVPTX/NVPTX.h:
```diff
@@ -106,15 +107,30 @@ enum LoadStore {
+  LAST = RelaxedMMIO
 };
+// Values match LLVM AtomicOrdering for common orderings:
+static_assert(Ordering::NotAtomic == (unsigned)AtomicOrdering::NotAtomic);
+static_assert(Ordering::Relaxed == (unsigned)AtomicOrdering::Monotonic);
+static_assert(Ordering::Acquire == (unsigned)AtomicOrdering::Acquire);
+static_assert(Ordering::Release == (unsigned)AtomicOrdering::Release);
+static_assert(Ordering::SequentiallyConsistent ==
+              (unsigned)AtomicOrdering::SequentiallyConsistent);
```
This current version does the job, but it still rubs me the wrong way, a bit. We're explicitly verifying that the values match those of  AtomicOrdering, and *implicitly* relying on the numeric order of enum values by hardcoding them as literals.

I think it would make more sense to explicitly assign enum values to their AtomicOrdering counterparts, and then *explicitly* enforce the ordering via the static_assert. This would better express our intent here.

On the other hand, it's convenient to see the explicit values all in one place, and mixing AtomicOrdering constants and explicit literals would probably look a bit odd, too.

I'll leave it up to you.

gonzalobg replied:
Not sure I fully understand what you are asking for yet. Is something like this what you have in mind:

```
enum Ordering : OrderingUnderlyingType {
  NotAtomic = AtomicOrdering::NotAtomic, // PTX calls these: "Weak"
  // Unordered = AtomicOrdering::Unordered, // NVPTX maps LLVM Unorderd to Relaxed
  Relaxed = AtomicOrdering::Monotonic,
  Acquire = AtomicOrdering::Acquire,
  Release = AtomicOrdering::Release,
  // AcquireRelease = AtomicOrdering::AcquireRelease,
  SequentiallyConsistent = AtomicOrdering::SequentiallyConsistent,
  Volatile = SequentiallyConsistent + 1,
  RelaxedMMIO = Volatile + 1,
  LAST = RelaxedMMIO
};
```

?

Artem-B replied:
Yeah, something like that. On a second thought, it's probably an overkill. While matching numeric values to AtomicOrdering is a convenience, it's not necessary. We're not extending AtomicOrdering, but, rather, defining an new ordering with some overlap with AtomicOrdering.

How important in practice is it to have the matching numeric values?

Perhaps what we need is a set of ordering values, w/o any attempt to inherit from AtomicOrdering, or comments that we map some values to something else, and, if needed, an explicit mapping function between AtomicOrdering and NVPTX ordering.

gonzalobg replied:
I think its mostly a convenience. When debugging, I think its nice if 0 means the same thing both in LLVM and in the NVPTX backend. This also allows casting the LLVM one into the NVPTX one (or just using either as a "machine code").

The LLVM values are explicitly set and don't change often (I don't think they've ever changed), so I am not too worried about how to do this (static cast like right now, vs explicitly setting the values to the LLVM enum values). But I do think that not having to remember that 3 means something in AtomicOrdering and something different in NVPTX ordering is worth it.

Artem-B commented on llvm/lib/Target/NVPTX/NVPTXUtilities.h:
```diff
@@ -82,6 +83,36 @@ inline unsigned promoteScalarArgumentSize(unsigned size) {
+}
+
+} // namespace NVPTX
 }
```
That's the end of `namespace llvm`. You may want to add a comment here.

Artem-B reviewed (approved):
LGTM. 
Let's wait till Monday to land it.

gonzalobg committed:
[NVPTX] Ordering uses AtomicOrdering values directly

mikaelholmen commented on llvm/lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:
```diff
@@ -1989,17 +1971,12 @@ bool NVPTXDAGToDAGISel::tryStore(SDNode *N) {
+    Ops.append({BasePtr, Chain});
   }
 
+  SDNode *NVPTXST = NVPTXST =
```
One assignment too much here?
gcc detects this and warns like
```
../lib/Target/NVPTX/NVPTXISelDAGToDAG.cpp:1977:29: warning: operation on 'NVPTXST' may be undefined [-Wsequence-point]
 1977 |   SDNode *NVPTXST = NVPTXST =
      |                     ~~~~~~~~^
 1978 |       CurDAG->getMachineNode(*Opcode, DL, MVT::Other, Ops);
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

gonzalobg replied:
Indeed, thank you for reporting this.
Fix here: https://github.com/llvm/llvm-project/pull/103045
